tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/firebase/FirebasePushClient.java

```diff
@@ -0,0 +1,69 @@
+package com.linagora.tmail.james.jmap.firebase;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.messaging.FirebaseMessaging;
+import com.google.firebase.messaging.Message;
+import com.linagora.tmail.james.jmap.model.FirebaseToken;
+import com.linagora.tmail.james.jmap.model.MissingOrInvalidFirebaseCredentialException;
+
+import reactor.core.publisher.Mono;
+
+public class FirebasePushClient {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FirebasePushClient.class);
+
+    private final FirebaseMessaging firebaseMessaging;
+
+    @Inject
+    public FirebasePushClient(FirebaseConfiguration configuration) {
+        try {
+            FirebaseOptions options = FirebaseOptions.builder()
+                .setCredentials(GoogleCredentials.fromStream(new FileInputStream(configuration.firebasePrivateKeyUrl())))
+                .build();
+
+            FirebaseApp.initializeApp(options);
+        } catch (FileNotFoundException e) {
+            LOGGER.error("Declared private key file for Firebase is not found");
+            throw new MissingOrInvalidFirebaseCredentialException(e.getMessage());
+        } catch (IOException e) {
+            LOGGER.error("Declared private key file for Firebase is invalid");
+            throw new MissingOrInvalidFirebaseCredentialException(e.getMessage());
+        }
+
+        this.firebaseMessaging = FirebaseMessaging.getInstance();
+    }
+
+    public Mono<Void> push(FirebasePushRequest pushRequest) {
+        return Mono.fromCallable(() -> firebaseMessaging.send(createFcmMessage(pushRequest)))
+            .then();
+    }
+
+    public Mono<Void> validateToken(FirebaseToken token) {
```

Comment:
Why do we need a separate call to validate tokens? 

Can't we just handle exceptions upon push and do a single call?

Reply:
> Can't we just handle exceptions upon push and do a single call?

Yes, we can, but should we not validate the token when user creates a FirebaseSubscription first?

Reply:
I'm not sure: sending a dummy message is not clean and do not matches the spec we wrote.
 
Is there a way to validate this without sending a dummy message ?
 
Also if that is needed then we should make that a first class citizen of our spec: "Upon establishing the Firebase connection the JMAP server sends an initial message at XYZ format (we could put the properties of the Firebase registration: `id, expires, types, devideId`)"... This would already be cleaner...

Reply:
Oh I saw `validateValidTokenShouldSucceedAndPushNoMessage`. So as the payload is empty firebase do not push it. Smart. If that is the case please dis-regard the previous comment.

Reply:
I guess Quan's idea is not only case when "empty payload". The Tmail will valid firebase token when client create a new subscription (and return error response for client as instantly).
for this case, I tend to validate token after (in listener).

Reply:
Actually, it is called `dryRun` and even with an actual message, FCM will not send the message either. But you got my idea.

Reply:
(+) It will more friendly for client, (client can ensure the token in the own request is correct, in this time).
(-) But, nothing ensure the token will be valid forever, so we still do again validation token (or handler exception) after. 

Just a option, I don't see a clear benefit between two option

Reply:
We can validate the token when user creates a subscription and handle exceptions also at push listener (say remove the subscription if not valid anymore exception) IMO

Reply:
When the token becomes no longer valid then at that moment we would remove the corresponding subscription.

Validating the token upon subscription is a good practice as it would ensures the user that the subscription he just created works.